server/channels/app/sso.go

```diff
@@ -0,0 +1,86 @@
+// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
+// See LICENSE.txt for license information.
+
+package app
+
+import (
+	"crypto/sha256"
+	"encoding/base64"
+	"encoding/hex"
+	"encoding/json"
+	"errors"
+	"net/http"
+
+	"github.com/mattermost/mattermost/server/public/model"
+)
+
+func (a *App) CreateCodeChallengeToken(codeChallenge string) (*model.Token, *model.AppError) {
+	extraProps, extraErr := json.Marshal(map[string]string{"code_challenge": codeChallenge})
+	if extraErr != nil {
+		return nil, model.NewAppError("App.CreateCodeChallengeToken", "api.oauth.store_code_challenge", nil, "", http.StatusInternalServerError).Wrap(extraErr)
+	}
+	token := model.NewToken("pkce", string(extraProps))
+	if err := a.Srv().Store().Token().Save(token); err != nil {
+		var appErr *model.AppError
+		switch {
+		case errors.As(err, &appErr):
+			return nil, model.NewAppError("App.CreateCodeChallengeToken", "api.oauth.store_code_challenge", nil, "", http.StatusInternalServerError).Wrap(appErr)
+		default:
+			return nil, model.NewAppError("App.CreateCodeChallengeToken", "api.oauth.store_code_challenge", nil, "", http.StatusInternalServerError).Wrap(err)
+		}
+	}
+	return token, nil
+}
+
+func (a *App) UpdateCodeChallengeToken(codeChallengeToken string, session *model.Session) *model.AppError {
+	token, err := a.Srv().Store().Token().GetByToken(codeChallengeToken)
```

Comment:
The token type must be checked before making any updates, otherwise it could be any kind of token